src/ext/mod.rs

```diff
@@ -179,22 +181,23 @@ impl HeaderCaseMap {
         self.0.get_all(name).into_iter()
     }
 
-    #[cfg(any(feature = "client", feature = "server"))]
-    pub(crate) fn default() -> Self {
-        Self(Default::default())
-    }
-
-    #[cfg(any(test, feature = "ffi"))]
-    pub(crate) fn insert(&mut self, name: HeaderName, orig: Bytes) {
+    /// Inserts a header spelling, replacing any existing ones associated with that header name.
+    #[cfg(all(any(feature = "client", feature = "server"), feature = "http1"))]
+    pub fn insert(&mut self, name: HeaderName, orig: Bytes) -> Result<(), InvalidHeaderName> {
+        HeaderName::from_bytes(&orig)?;
```

Comment:
Clever! Hm, at the same time, it worries me a little to be [validating, not parsing](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/). It does mean a user can't round trip from `get_all` to `insert` without having to validate again.

At the same time, is it really worth having a separate `CasedHeaderName` type? Maybe it is... This is probably the hardest question with making this public.

Also, I think `HeaderName::from_bytes()` will allocate a new copy... That could probably be handled with `from_maybe_shared`, but just wanted to note for myself at least.

Reply:
[HeaderName](https://docs.rs/http/latest/http/header/struct.HeaderName.html) doesn't seem to have `from_maybe_shared`, that seems to be exclusive to [HeaderValue](https://docs.rs/http/latest/http/header/struct.HeaderValue.html)?

Anyway, I added a CasedHeaderName type. Usage is a bit clunky since its used like `.insert(CasedHeaderName(name, orig))`.
There's also a `From<Bytes>` implementation that infers the header name by parsing it.

Reply:
Oops, forgot to enforce the parsing of the header name, fixing now...